### PR TITLE
JAMES-4025 Rewrite some tests of JMAP draft on top of JMAP RFC-8621

### PR DESCRIPTION
**Pull request by Arsnael**
3 more to go...

### CODE REVIEW

**Review by chibenwa — COMMENTED** [review]
:+1:

**Comment by Arsnael** [comment]
There is different behavior I think between jmap draft and rfc and I aint sure if on rfc some are wanted (because spec just changed?) or not. Couldnt find anything clear in both specs either (or maybe my head is just too deep in it).

Last commit:
- first disabled test: we don't seem to do any intersect if the flag is draft, we only do a union of flags every time it seems on a same message inn different mailboxes. Different from draft I guess? (didnt dig in draft code but if it works on the test then I guess it does...). We missing something here, or with final jmap rfc we have just no intersection for such a case?
- second disabled test, when updating mail via email/set method jmap rfc, we seem to merge all flags together from both metadata as originFlags, and then compared to newFlags. If it's the same (which is the case here?) we dont do any update, thus not solving the inconsistency. We should find a way to fix that correct? Well if I understand correctly what is going on.

cc @chibenwa

**Comment by Arsnael** [comment]
After talking with @chibenwa 
- first case: drop it, union on flags, no need an complex intersect just for draft, not necessary
- second case: investigate code on draft, see what's made different. We should be able to fix the inconsistency with a email/set flag so potentially fix the code

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ImapKeywordsConsistencyContract.scala [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+    // When user asks for message list in mailbox "inbox" with flag "$Flagged"
+    val ids = listMessageIdsByMailboxAndKeyword(server, DefaultMailboxes.INBOX, "$Flagged").asScala.toList
+
+    // Then the message list is empty
+    assertThat(ids.size).isEqualTo(0)
+  }
+
+  @Disabled("TODO: Fix!")
+  @Test
+  def emailSetShouldSucceedToSolveKeywordsConflictsIntroducedViaImapUponFlagsAddition(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ImapKeywordsConsistencyContract.scala:241]: That one is a tricky one, tried to solve it quickly but every time it created problems elsewhere... 

I think it's better to leave it disabled for now, and fix it in a following PR

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ImapKeywordsConsistencyContract.scala:241]: Created a task for it => https://github.com/linagora/james-project/issues/5158

This PR is ready for review

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ImapKeywordsConsistencyContract.scala:241]: Please create an ASF task and put ASF ticket number into the @Disabled description

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ImapKeywordsConsistencyContract.scala:241]: JAMES-4026

**Comment by Arsnael** [comment]
Note: Need to rework this, as some PRs moving some resources (like JmapGuiceProbe) been moved

**Comment by Arsnael** [comment]
Rebased and fixed imports on JmapGuiceProbe

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationIntegrationTest.scala [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+  @Test
+  def jmapVacationShouldGenerateAReplyWhenActive(server: GuiceJamesServer): Unit = {
+    /* Test scenario :
+      - bob sets a Vacation on its account
+      - andre matthieu@mydomain.tld sends bob a mail
+      - bob should well receive this mail
+      - andre should well receive a notification about bob vacation
+    */
+
+    val mailboxProbe = server.getProbe(classOf[MailboxProbeImpl])
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationIntegrationTest.scala:97]: we can add the code bellow in `@BeforeEach`
I see it repeat in each test

```
    val mailboxProbe = server.getProbe(classOf[MailboxProbeImpl])
    val bobInboxId = mailboxProbe.createMailbox(MailboxConstants.USER_NAMESPACE, BOB.asString, DefaultMailboxes.INBOX)
    val andreInboxId = mailboxProbe.createMailbox(MailboxConstants.USER_NAMESPACE, ANDRE.asString, DefaultMailboxes.INBOX)
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaMailingTest.scala [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when
+      .post()
+    .`then`
+      .statusCode(SC_OK)
+      .log.ifValidationFails()
+      .body("methodResponses[0][1].list.preview",
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaMailingTest.scala:158]: Why duplicate check `"methodResponses[0][1].list.preview"` here?
which one is correct?

**Review by quantranhong1999 — APPROVED** [review]
LGTM

wait for a green build.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/CorsHeaderAPITest.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+               |    {
+               |      "accountId": "$ACCOUNT_ID"
+               |    },
+               |    "c1"]]
+               |}""".stripMargin)
+    .when
+      .post
+    .`then`
+      .header("Access-Control-Allow-Origin", "*")
+      .header("Access-Control-Allow-Methods", "GET, POST, DELETE, PUT")
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/CorsHeaderAPITest.scala:64]: Do you think we should update `Access-Control-Allow-Methods` in jmap server?
I think we just allows:
- POST: Jmap, upload
- GET: download
- OPTIONS: for check cors
No any use PUT and DELETE

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationIntegrationTest.scala (hunk 2 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationIntegrationTest.scala:97]: I chose not to. For the simple reason that I need the mailbox ids and with scala if I add vars in the class for that, it asks me to override and implement them in the memory and distributed class of the contract, which I found particularly annoying and unecessary

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/CorsHeaderAPITest.scala (hunk 4 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/CorsHeaderAPITest.scala:64]: If you think it's worth the reflection, open a ticket, but not the concern of this work here IMO :)

Comment (vttranlina) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/CorsHeaderAPITest.scala:64]: I created a ticket for it https://github.com/linagora/james-project/issues/5159

**Review by hungphan227 — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaMailingTest.scala (hunk 3 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaMailingTest.scala:158]: Both... it checks items in both previews (not sure can predict the return order?), checking that in the list we got both messages

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaMailingTest.scala:158]: That's more or less what the draft test is doing too

Comment (vttranlina) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaMailingTest.scala:158]: ah, understood

**Review by vttranlina — APPROVED** [review]

**Comment by Arsnael** [comment]
Capricious CI...

**Pull request merged**